File: packages/gavio-js/src/interceptors/governance/cost-router.ts
/** costRouter (F-GOV-06) — auto-route simple prompts to a cheaper model. */

import type { InterceptorContext } from '../../context.js'
import { estimateTokens } from '../../pricing.js'
import type { GavioRequest } from '../../request.js'
import type { Interceptor } from '../base.js'

const REASONING_KEYWORDS = new Set([
  'why', 'because', 'compare', 'trade-off', 'tradeoff', 'explain', 'analyze',
  'analyse', 'evaluate', 'design', 'architecture', 'review', 'debug',
  'reasoning', 'justify', 'critique',
])
const TOKEN = /[a-z0-9-]+/g

/** Scores prompt text in `[0, 1]` — higher means more complex. */
export interface ComplexityScorer {
  score(text: string): number
}

/** Zero-dependency default: prompt length + reasoning-keyword density. */
export function heuristicComplexityScorer(): ComplexityScorer {
  return {
    score(text: string): number {
      const tokens = estimateTokens(text)
      const lengthScore = Math.min(tokens / 200, 1.0) * 0.6
      const words = new Set(text.toLowerCase().match(TOKEN) ?? [])
      let keywordHits = 0
      for (const w of words) if (REASONING_KEYWORDS.has(w)) keywordHits += 1
      const keywordScore = Math.min(keywordHits / 3, 1.0) * 0.4
      return Math.min(lengthScore + keywordScore, 1.0)
    },
  }
}

export interface CostRouterOptions {
  simpleModel: string
  complexityThreshold?: number
  scorer?: ComplexityScorer
}

/**
 * Reroute a request to `simpleModel` when its complexity score is low.
 *
 * Register early in the chain, before caching, so a rerouted request's cache
 * key reflects the model it actually ran on. Register after `modelPolicy` if
 * RBAC should gate on the caller's *requested* model, not the rerouted one.
 */
export function costRouter(options: CostRouterOptions): Interceptor {
  const { simpleModel, complexityThreshold = 0.35, scorer = heuristicComplexityScorer() } = options
  return {
    name: 'cost_router',
    before(request: GavioRequest, ctx: InterceptorContext): GavioRequest {
      const score = scorer.score(request.promptText())
      const rerouted = score < complexityThreshold && request.model !== simpleModel
      ctx.state['cost_router'] = {
        rerouted,
        originalModel: request.model,
        complexityScore: score,
      }
      if (!rerouted) return request
      const result = request.copyWithMessages(request.messages)
      result.model = simpleModel
      return result
    },
  }
}

File: packages/gavio-py/gavio/interceptors/governance/cost_router.py
"""CostRouter (F-GOV-06) — auto-route simple prompts to a cheaper model."""

from __future__ import annotations

import re
from typing import Protocol, runtime_checkable

from ...context import InterceptorContext
from ...pricing import estimate_tokens
from ...request import GavioRequest
from ..base import Interceptor

_REASONING_KEYWORDS = frozenset(
    {
        "why", "because", "compare", "trade-off", "tradeoff", "explain",
        "analyze", "analyse", "evaluate", "design", "architecture", "review",
        "debug", "reasoning", "justify", "critique",
    }
)
_TOKEN = re.compile(r"[a-z0-9-]+")


@runtime_checkable
class ComplexityScorer(Protocol):
    """Scores prompt text in ``[0, 1]`` — higher means more complex."""

    def score(self, text: str) -> float: ...


class HeuristicComplexityScorer:
    """Zero-dependency default: prompt length + reasoning-keyword density."""

    def score(self, text: str) -> float:
        tokens = estimate_tokens(text)
        length_score = min(tokens / 200, 1.0) * 0.6
        words = set(_TOKEN.findall(text.lower()))
        keyword_hits = len(words & _REASONING_KEYWORDS)
        keyword_score = min(keyword_hits / 3, 1.0) * 0.4
        return min(length_score + keyword_score, 1.0)


class CostRouter(Interceptor):
    """Reroute a request to ``simple_model`` when its complexity score is low.

    Register early in the chain, before caching, so a rerouted request's cache
    key reflects the model it actually ran on. Register after ``ModelPolicy``
    if RBAC should gate on the caller's *requested* model, not the rerouted one.
    """

    def __init__(
        self,
        simple_model: str,
        complexity_threshold: float = 0.35,
        scorer: ComplexityScorer | None = None,
    ) -> None:
        self.simple_model = simple_model
        self.complexity_threshold = complexity_threshold
        self.scorer = scorer or HeuristicComplexityScorer()

    @property
    def name(self) -> str:
        return "cost_router"

    async def before(
        self, request: GavioRequest, ctx: InterceptorContext
    ) -> GavioRequest:
        score = self.scorer.score(request.prompt_text())
        rerouted = score < self.complexity_threshold and request.model != self.simple_model
        ctx.state["cost_router"] = {
            "rerouted": rerouted,
            "original_model": request.model,
            "complexity_score": score,
        }
        if not rerouted:
            return request
        result = request.copy_with_messages(request.messages)
        result.model = self.simple_model
        return result
